Add aws_mfa_devices module for multi-factor authentication on AWS

pwnall opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aws_mfa_devices

##### ANSIBLE VERSION
```
ansible 2.0.1.0
```

##### SUMMARY
Amazon's [GetSessionToken API](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html), as well as the 
[AssumeRole API](http://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) behind the [sts_assume_role module](http://docs.ansible.com/ansible/sts_assume_role_module.html) require a MFA token serial number.

This module can be used to obtain the serial numbers for all the user's MFA devices, and can be combined with a `prompt` to obtain all the information for the APIs listed above.

I tested my code with the following rules:

```
- name: list MFA devices
  aws_mfa_devices:
    region: "{{ aws_region }}"
  register: mfa_devices

- debug: var=mfa_devices
```

Output:
```
ok: [localhost] => {
    "mfa_devices": {
        "changed": true,
        "devices": [
            {
                "enable_date": "2016-03-11T23:25:36+00:00",
                "serial_number": "arn:aws:iam::085120003701:mfa/victor_costan",
                "user_name": "victor_costan"
            }
        ]
    }
}
```

gregdek commented:
Thanks @pwnall for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

pwnall commented:
ready_for_review

pwnall commented:
'ready_for_review'

gregdek commented:
Thanks @pwnall for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

dorianj commented:
shipit -- I've been using this locally and it works for me.

gregdek commented:
Thanks @pwnall and @dorianj -- will get this into the shipit pipeline.

gregdek commented:
Thanks @pwnall for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

pwnall commented:
@gregdek PTAL?

pwnall commented:
@gregdek This PR passed CI and has received a [shipit endorsement](https://github.com/ansible/ansible-modules-extras/pull/1988#issuecomment-212208059). The only change I had to make since then involved bumping `version_added`. Can you please take another look?

gregdek commented:
@pwnall will put into the agenda for our upcoming new module review meeting.

gundalow commented on cloud/amazon/aws_mfa_devices.py:
```diff
@@ -0,0 +1,133 @@
+    - aws
+    - ec2
+requirements:
+    - boto3
```
Can you please add `botocore` to this list

gundalow commented on cloud/amazon/aws_mfa_devices.py:
```diff
@@ -0,0 +1,133 @@
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+      module.fail_json(msg='boto3 is required.')
```
`boto3 and botocore are required.`

gundalow replied:
Could you please check the indentation here, looks like you are missing two spaces. Thanks

gundalow replied:
Can you please change this to `boto3 and botocore are required.`

gundalow commented on cloud/amazon/aws_mfa_devices.py:
```diff
@@ -0,0 +1,133 @@
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
+
+main()
```
This needs to be as follows so debuggers can work with ziploader:

```
if __name__ == '__main__':
    main()
```

gundalow commented:
needs_revision Nothing major. The previous endorsements should still count once my comments have been addressed.

pwnall commented:
@gundalow Can you please take another look? I hope I've addressed your feedback.

gundalow commented on cloud/amazon/aws_mfa_devices.py:
```diff
@@ -0,0 +1,135 @@
+DOCUMENTATION = '''
+---
+module: aws_mfa_devices
+short_description: List the MFA devices registered for a user
```
When skimming the list of modules I think this will make it clearer:

`List the MFA (Multi-Factor Authentication) devices registered for a user`

gundalow commented on cloud/amazon/aws_mfa_devices.py:
```diff
@@ -0,0 +1,135 @@
+module: aws_mfa_devices
+short_description: List the MFA devices registered for a user
+description:
+    - List the MFA devices registered for a user
```
`List the MFA (Multi-Factor Authentication)devices registered for a user.`

gundalow commented:
needs_revision

gundalow commented:
@pwnall Thanks for this module, I didn't get a chance to do a full review before.

Once my comments have been addressed I'm happy that this meets http://docs.ansible.com/ansible/developing_modules.html#module-checklist and https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/GUIDELINES.md

Given @dorianj has said "This works for me", once my comments have been addressed we can get this merged.

pwnall commented:
@gundalow Thank you very much for the feedback, and for your work on Ansible!

I think I've addressed all the feedback items, and I'd be happy to do anything else I can to help this get merged!

gundalow commented:
@pwnall Thank you for the quick turn around on this, and your contribution to Ansible

shipit

_**Ready for merge**_

**Details of review**
- @gundalow has reviewed against  http://docs.ansible.com/ansible/developing_modules.html#module-checklist
  - @gundalow has reviewed against  https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/GUIDELINES.md
- @dorianj stated:  _"I've been using this locally and it works for me."_

gregdek commented:
Thanks again to @pwnall for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

abadger commented:
Merged to devel

pwnall commented:
@gundalow Once again, thank you very much for all your help! I'm really grateful for the time you spent reviewing my work and helping me improve! ❤️

gundalow commented:
@pwnall Thank you for the new module. Hope to see your name around here if you have more good stuff to fix :)
